packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java

```diff
@@ -0,0 +1,332 @@
+              if (android.os.Build.VERSION.SDK_INT < 26) {
+                promise.resolve(Build.SERIAL);
+              } else {
+                promise.resolve(Build.getSerial());
+              }
+            } catch (SecurityException e) {
+              Log.e(TAG, e.getMessage());
+              promise.reject(e);
+            }
+          } else {
+            promise.reject(new SecurityException("User rejected permissions"));
+          }
+        }
+      });
+    }
+  }
+
+  @ExpoMethod
+  public void getFreeDiskStorageAsync(Promise promise) {
+    try {
+      StatFs external = new StatFs(Environment.getExternalStorageDirectory().getAbsolutePath());
+      long availableBlocks;
+      long blockSize;
+
+      availableBlocks = external.getAvailableBlocksLong();
+      blockSize = external.getBlockSizeLong();
+
+      BigInteger storage = BigInteger.valueOf(availableBlocks).multiply(BigInteger.valueOf(blockSize));
+      promise.resolve(storage.doubleValue());
+    } catch (NullPointerException e) {
+      Log.e(TAG, e.getMessage());
+      promise.reject(e);
+    }
+  }
+
+  @ExpoMethod
+  public void getIpAddressAsync(Promise promise) {
+    try {
+      Integer ipAddress = getWifiInfo().getIpAddress();
+      // Convert little-endian to big-endianif needed
+      if (ByteOrder.nativeOrder().equals(ByteOrder.LITTLE_ENDIAN)) {
+        ipAddress = Integer.reverseBytes(ipAddress);
+      }
+      byte[] ipByteArray = BigInteger.valueOf(ipAddress).toByteArray();
+      String ipAddressString = InetAddress.getByAddress(ipByteArray).getHostAddress();
+      promise.resolve(ipAddressString);
+    } catch (Exception e) {
+      Log.e(TAG, e.getMessage());
+      promise.reject(e);
+    }
+  }
+
+  @ExpoMethod
+  public void getMACAddressAsync(String interfaceName, Promise promise) {
+    String permission = "android.permission.INTERNET";
+    int res = mContext.checkCallingOrSelfPermission(permission);
+
+    String macAddress = "";
+    if (res == PackageManager.PERMISSION_GRANTED) {
+      try {
+        List<NetworkInterface> interfaces = Collections.list(NetworkInterface.getNetworkInterfaces());
+        for (NetworkInterface intf : interfaces) {
+          if (interfaceName != null) {
+            if (!intf.getName().equalsIgnoreCase(interfaceName)) continue;
+          }
+          byte[] mac = intf.getHardwareAddress();
+          if (mac == null) {
+            macAddress = "";
+          }
+          StringBuilder buf = new StringBuilder();
+          for (byte aMac : mac) {
+            buf.append(String.format("%02X:", aMac));
+          }
+          if (buf.length() > 0) {
+            buf.deleteCharAt(buf.length() - 1);
+          }
+          macAddress = buf.toString();
+          break;
+        }
+      } catch (Exception e) {
+        Log.e(TAG, e.getMessage());
+        promise.reject(e);
+      }
+    } else {
+      promise.reject(new AccessControlException("No permission granted to access the Internet"));
```

Comment:
same as above - no need to create a new AccessControlException, instead just pass in an error code and message directly to promise.reject.